compiler/rustc_mir_dataflow/src/impls/single_enum_variant.rs

```diff
@@ -0,0 +1,171 @@
+
+    const NAME: &'static str = "single_enum_variant";
+
+    fn bottom_value(&self, _: &mir::Body<'tcx>) -> Self::Domain {
+        //FactArray { arr: [Fact::TOP; 128] }
+        FactCache::new(Local::from_u32(0), Location::START, VariantIdx::MAX)
+    }
+
+    fn initialize_start_block(&self, body: &mir::Body<'tcx>, state: &mut Self::Domain) {
+        // assume everything is top initially.
+        let local_decls = body.local_decls();
+        for (l, _) in local_decls.iter_enumerated() {
+            state.remove(l);
+            //state.insert(l, Fact::TOP);
+            //state[l] = Fact::TOP;
+        }
+    }
+}
+
+impl<'tcx> SingleEnumVariant<'_, 'tcx> {
+    pub fn new<'a>(tcx: TyCtxt<'tcx>, body: &'a mir::Body<'tcx>) -> SingleEnumVariant<'a, 'tcx> {
+        SingleEnumVariant { tcx, body }
+    }
+    #[inline]
+    pub fn is_tracked(&self, place: &Place<'tcx>) -> bool {
+        place.ty(self.body, self.tcx).ty.is_enum()
+    }
+    fn assign(
+        &self,
+        state: &mut <Self as AnalysisDomain<'tcx>>::Domain,
+        lhs: &Place<'tcx>,
+        rhs: &Operand<'tcx>,
+        location: Location,
+    ) {
+        let _: Option<_> = try {
+            if !self.is_tracked(lhs) {
+                return;
+            }
+            let lhs_local = lhs.local_or_deref_local()?;
+
+            let new_fact = match rhs {
+                Operand::Copy(rhs) | Operand::Move(rhs) => {
+                    if let Some(rhs_local) = rhs.local_or_deref_local() {
+                        state.get(rhs_local).map(|f| f.1).copied()
+                    } else {
+                        rhs.ty(self.body, self.tcx).variant_index.map(|var_idx| var_idx)
+                    }
+                }
+                // Assigning a constant does not affect discriminant?
+                Operand::Constant(_c) => return,
+            };
+            if let Some(new_fact) = new_fact {
+                state.insert(lhs_local, location, new_fact);
+            } else {
+                state.remove(lhs_local);
+            }
+        };
+    }
+}
+
+impl<'tcx> Analysis<'tcx> for SingleEnumVariant<'_, 'tcx> {
+    fn apply_statement_effect(
+        &self,
+        state: &mut Self::Domain,
+        statement: &Statement<'tcx>,
+        loc: Location,
+    ) {
+        let (place, fact) = match &statement.kind {
+            StatementKind::Deinit(box place) => (place, None),
+            StatementKind::SetDiscriminant { box place, variant_index } => {
+                (place, Some(*variant_index))
+            }
+            StatementKind::Assign(box (lhs, Rvalue::Use(op))) => {
+                return self.assign(state, lhs, op, loc);
+            }
+            /* may alias/mutate RHS need to specify that it is no longer a single value */
+            StatementKind::Assign(box (
+                _,
+                Rvalue::Ref(_, BorrowKind::Mut { .. }, rhs)
+                | Rvalue::AddressOf(Mutability::Mut, rhs),
+            )) => (rhs, None),
```

Comment:
Unfortunately, a reference/pointer to a field of an enum (eg `(_5 as Some).0` can be used to change the "active variant" of the enum in niche optimized cases. This is discussed in rust-lang/unsafe-code-guidelines#84 . I've adjusted Ralf's example from there a little bit to this:

```rust
fn main() {
    dbg!(foo());
}

#[inline(never)]
fn foo() -> u32 {
    unsafe {
        let mut x = Some(&0);
        match x {
            Some(ref mut b) => {
                let u = b as *mut &i32 as *mut usize;
                // Just writing into a *mut u8
                *u = 0;
            }
            None => {
                return 100;
            }
        }
        match x {
            Some(_) => 1,
            None => 2,
        }
    }
}
```

Looking at the produced MIR, I thought this should miscompile - however, it doesn't. I believe that's because the first match actually gets optimized and so the fact cache throws the information out before the second match. I'm not sure if it's actually possible to reproduce this miscompilation then, but still, this should be fixed. The past path towards a fix is going to depend on how some of the other issues are addressed though, so it should be fine to leave this one for later.

Reply:
I would think it would compile correctly because it just throws out the information when it sees a reference, and doesn't just ignore it?

Reply:
The problem is that the reference is of the form `&mut (_5 as Some).0`, which means the `local_or_deref_local` will return `None`, and so you won't get to the point where you throw the information out